daemon/oci_linux.go

```diff
@@ -748,12 +747,9 @@ func WithCommonOptions(daemon *Daemon, c *container.Container) coci.SpecOpts {
 			if (c.HostConfig.Init != nil && *c.HostConfig.Init) ||
 				(c.HostConfig.Init == nil && daemon.configStore.Init) {
 				s.Process.Args = append([]string{inContainerInitPath, "--", c.Path}, c.Args...)
-				path := daemon.configStore.InitPath
-				if path == "" {
-					path, err = exec.LookPath(dconfig.DefaultInitBinary)
-					if err != nil {
-						return err
-					}
+				path, err := daemon.configStore.LookupInitPath() // this will fall back to DefaultInitBinary and return an absolute path
```

Comment:
I'm wondering; would this be something we should do when initialising the daemon so that we can fail starting if the init-binary was not found?

Reply:
(Basically; it looks like currently we're assuming the happy path (`daemon start`, `container create`) all the way until the very last moment; when we create the OCI spec (`container start`) 😬

Reply:
TBH I was Today Years Old that I realised we do a `exec.LookPath()` for every container. Which also brings the interesting bit that `docker info` may show information about a different `docker-init` than that will be used by the container 🤔 (if a `docker-init` binary was dropped somewhere in `PATH`).

My thinking was mostly to resolve the binary once, and use that for the duration of the daemon's lifetime (or config reload if we decide to support that).
That would also allow us to show the location of the binary that will be used (instead of "you don't know what you're gonna get").

Reply:
`docker info` may show information about a different `docker-init` than what will be used by a container even if the path remains constant. The binary could be updated in between the `docker info` and `docker run`, for example. `docker info` returns a point-in-time snapshot of the daemon state, _and that's okay._ Unless we can guarantee that the `docker-init` binary used when starting containers can never ever change for the lifetime of a daemon process, I don't think we should do half measures like memoizing the resolved path to the binary for the duration of the daemon process. We shouldn't make promises we can't keep.